cargo: point to rust-driver 64b4afcd

muzarski opened the pull request:
Currently, there are some open PRs that depend on changes recently introduced to rust-driver.
Let's bump the version of rust-driver dependency.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] PR description sums up the changes and reasons why they should be introduced.
- ~[ ] I have implemented Rust unit tests for the features/changes introduced.~
- ~[ ] I have enabled appropriate tests in `.github/workflows/build.yml` in `gtest_filter`.~
- ~[ ] I have enabled appropriate tests in `.github/workflows/cassandra.yml` in `gtest_filter`.~

Lorak-mmk commented on scylla-rust-wrapper/src/query_result.rs:
```diff
@@ -43,11 +43,24 @@ impl CassResultData {
             )
         });
 
+        // col_specs received from QueryResult are borrowed.
+        // Need to convert them to owned version.
+        let col_specs = col_specs
+            .iter()
+            .map(|col_spec| {
+                // TODO: introduce ColumnSpec::[in]to_owned() methods in rust-driver.
+                let name = col_spec.name().to_owned();
+                let typ = col_spec.typ().clone().into_owned();
+                let table_spec = col_spec.table_spec().to_owned();
+                ColumnSpec::owned(name, typ, table_spec)
+            })
+            .collect();
+
```
When is this called? On each query?

wprzytula replied:
Yes, unfortunately. But soon it will be irrelevant, when we switch to new `QueryResult` with its `RowDeserializerOwned`.

muzarski replied:
Yeah... It's because current `QueryResult` API is in awkward state.

The main issue is that ColumnSpec lifetime is bound by lifetime of result:
```rust
    /// Returns column specifications.
    #[inline]
    pub fn col_specs(&self) -> &[ColumnSpec] {
        self.metadata
            .as_ref()
            .map(|metadata| metadata.col_specs())
            .unwrap_or_default()
    }
```

This is equivalent to:
```rust
pub fn col_specs<'s>(&'s self) -> &'s [ColumnSpec<'s>]
```
even though this version of `QueryResult` holds metadata consisting of `ColumnSpec<'static>` (owned column specs). To return `&[ColumnSpec<'static>]`, we also need fix introduced by this commit: https://github.com/scylladb/scylla-rust-driver/pull/1093/commits/e53a92c1721738dce4a6a4d257e5e0e48edaee21

Then we wouldn't need to call this `map` in cpp-rust-driver:
```rust
  .map(|col_spec| {
      // TODO: introduce ColumnSpec::[in]to_owned() methods in rust-driver.
      let name = col_spec.name().to_owned();
      let typ = col_spec.typ().clone().into_owned();
      let table_spec = col_spec.table_spec().to_owned();
      ColumnSpec::owned(name, typ, table_spec)
  })
```

I could open a small PR in rust-driver that does that (so we get rid of this mapping from borrowed col spec to owned col spec). This would be a temporary fix, before the final version of `QueryResult` is introduced in rust-driver.

As a side note: I believe that once final version of lazy deserialization is merged, the corresponding structures in cpp-rust-driver will need to be adjusted/refactored appropriately so we are as efficient as we can.

wprzytula commented on scylla-rust-wrapper/src/query_result.rs:
```diff
@@ -43,11 +43,24 @@ impl CassResultData {
             )
         });
 
+        // col_specs received from QueryResult are borrowed.
+        // Need to convert them to owned version.
+        let col_specs = col_specs
+            .iter()
+            .map(|col_spec| {
+                // TODO: introduce ColumnSpec::[in]to_owned() methods in rust-driver.
+                let name = col_spec.name().to_owned();
+                let typ = col_spec.typ().clone().into_owned();
+                let table_spec = col_spec.table_spec().to_owned();
+                ColumnSpec::owned(name, typ, table_spec)
+            })
+            .collect();
```
I haven't included `ColumnSpec::into_owned` on purpose: it would be inefficient, as another Vec would have to be allocated.
This shouldn't be needed; the `ColumnSpec` stored in `QueryResult` (soon `LegacyQueryResult`) is `'static`.
Actually, there is a bug in `QueryResult`:
```rust
pub fn col_specs(&self) -> &[ColumnSpec] {
```

...while it should be:
```rust
pub fn col_specs(&self) -> &[ColumnSpec<'static>] {
```

muzarski replied:
> Actually, there is a bug in `QueryResult`:
> 
> ```rust
> pub fn col_specs(&self) -> &[ColumnSpec] {
> ```
> 
> ...while it should be:
> 
> ```rust
> pub fn col_specs(&self) -> &[ColumnSpec<'static>] {
> ```

Yes, that's what I mentioned in response to @Lorak-mmk 's comment.

muzarski commented on scylla-rust-wrapper/src/query_result.rs:
```diff
@@ -23,15 +23,15 @@ pub struct CassResult {
     pub fn from_result_payload(
         paging_state_response: PagingStateResponse,
-        col_specs: Vec<ColumnSpec>,
+        col_specs: &[ColumnSpec],
```
This also needs adjustment in `QueryResult::col_specs()`, so it returns `&[ColumnSpec<'static>]`

muzarski commented:
v1.1: In rust-driver, we changed an error message emitted when statement id changes after repreparation. I adjusted the corresponding intergration test (`PreparedTests.Integration_Cassandra_FailFastWhenPreparedIDChangesDuringReprepare`) to this new error message.

muzarski commented:
v2.1: Bumped again to https://github.com/scylladb/scylla-rust-driver/commit/64b4afcdb4286b21f6cc1acb55266d6607f250e0

Lorak-mmk commented on scylla-rust-wrapper/src/cass_types.rs:
```diff
@@ -422,7 +422,7 @@ impl CassDataType {
 
 pub fn get_column_type(column_type: &ColumnType) -> CassDataType {
     match column_type {
-        ColumnType::Custom(s) => CassDataType::Custom((*s).clone()),
+        ColumnType::Custom(s) => CassDataType::Custom(s.clone().into_owned()),
         ColumnType::Ascii => CassDataType::Value(CassValueType::CASS_VALUE_TYPE_ASCII),
```
`s` is a `&String`, right? So why do we need both `.clone()` and `.into_owned()`? Isn't one of those enough?

muzarski replied:
`s` is `&Cow<'_, str>`. We need to `clone` the `Cow` and then convert it to `String`.

muzarski replied:
`str::to_owned(s)` or `s.to_string()` seems to work as well. In case you prefer something shorter compared to current version. Notice that `s.to_owned()` returns an owned `Cow` (not `String`).

Lorak-mmk replied:
> `s` is `&Cow<'_, str>`. We need to `clone` the `Cow` and then convert it to `String`.

I didn't realize that. In that case the current version is fine.

Lorak-mmk commented on scylla-rust-wrapper/src/cass_types.rs:
```diff
@@ -459,10 +459,15 @@ pub fn get_column_type(column_type: &ColumnType) -> CassDataType {
             field_types: field_types
                 .iter()
-                .map(|(name, col_type)| ((*name).clone(), Arc::new(get_column_type(col_type))))
+                .map(|(name, col_type)| {
+                    (
+                        name.clone().into_owned(),
+                        Arc::new(get_column_type(col_type)),
+                    )
+                })
```
ditto

Lorak-mmk commented on scylla-rust-wrapper/src/cass_types.rs:
```diff
@@ -459,10 +459,15 @@ pub fn get_column_type(column_type: &ColumnType) -> CassDataType {
                 .collect(),
-            keyspace: (*keyspace).clone(),
-            name: (*type_name).clone(),
+            keyspace: keyspace.clone().into_owned(),
+            name: type_name.clone().into_owned(),
```
ditto

Lorak-mmk commented on scylla-rust-wrapper/src/exec_profile.rs:
```diff
@@ -376,13 +376,13 @@ pub unsafe extern "C" fn cass_execution_profile_set_retry_policy(
-        DefaultRetryPolicy(default) => default.as_ref(),
-        FallthroughRetryPolicy(fallthrough) => fallthrough.as_ref(),
-        DowngradingConsistencyRetryPolicy(downgrading) => downgrading.as_ref(),
+    let retry_policy: Arc<dyn RetryPolicy> = match ptr_to_ref(retry_policy) {
+        DefaultRetryPolicy(default) => default.clone(),
+        FallthroughRetryPolicy(fallthrough) => fallthrough.clone(),
+        DowngradingConsistencyRetryPolicy(downgrading) => downgrading.clone(),
```
Please use `Arc::clone` if you need to clone `Arc`. `.clone` is less obvious.
